src/pip/_internal/req/req_install.py

```diff
@@ -436,6 +436,7 @@ def _correct_build_location(self):
         This is only called by self.egg_info_path to fix the temporary build
         directory.
         """
+        self._egg_info_path = None
```

Comment:
Instead of setting `self._egg_info_path` explicitly everywhere, can we turn `source_dir` into a property and then set `self._egg_info_path = None` in the setter?